### PR TITLE
tlb: support empty metadata keys

### PR DESCRIPTION
**Pull request by doudou**
The metadata system makes a difference between a key that is set to
an empty set of values, and a non-existent key. The TLB import/export
was losing this distinction.

### CODE REVIEW

**Comment by marvin2k** [comment]
Hm. Now `drivers/orogen/dynamixel` is successfully building with the clang-importer (from [here](https://github.com/orocos-toolchain/orogen/pull/63), the clang importer is not calling `Typelib::Registry.add_standard_cxx_types(registry)`). But the fix is counter-intuitive: How does the gccxml importer circumvents the limitation fixed in this PR by calling `add_standard_cxx_types`? The resulting entries have no metadata at all... Some aliasing?

Then, as a follow-up: Is this `add_standard_cxx_types` feature something to keep? In the `clang` branch it is (after a quick glance) only used for testing... What about [standard_conversions.rb](https://github.com/orocos-toolchain/typelib/blob/master/bindings/ruby/lib/typelib/standard_convertions.rb). All this does not make sense: either typelib can work with a given set of types without any support (looking at `base/orogen/std`) or this support is coming from external. But having two different sources for information about "/int32_t" or "/std/string"... this is bound to break...

**Comment by doudou** [comment]
The gccxml importer currently does not import the fundamental types (I think ... maybe it does). The addition of add_standard_cxx_types was working around the bug as packages that generate type-related code would have a typekit, would therefore get a registry where *in process* the standard set of types have been added and normalized by #resolve_registry_includes (which is what base/orogen/std does manually).

Typelib does not care about any set of types, and it's not the importer's job to define what the "orogen fundamental set of types" is. So, indeed, add_standard_cxx_types does not really have its place anymore in typelib, but should be replaced by base/orogen/std (which is nowadays meant to define "the fundamental set of types for orogen-generated RTT systems").

As I stated in the clang discussions, the cleanest would be to have the C++ importers do purely C++ import and orogen "translate" the types to the "orogen fundamental set of types". It would keep the C++ importer free of the orogen conventions (using uintXX and so on). In the long run, I'd like to even move on the orogen side to something closer to the ROS convention (no _t and float32/float64 for float and double).

The use of add_standard_cxx_types in standard_convertions.rb is due to the lack of a Character type class in typelib, thus using a numeric type to represent chars. Since the signed-ness of 'char' is architecture-dependent, it had to be "discovered". That's 11 years of having-other-more-urgent-things-to-fix for you.

TL;DR: the way forward here is:
 - add a Character type class in typelib, use it for char and wchar_t (and /std/string)
 - make sure the gccxml importer and clang deal only with standard C++ types
 - translate the C++ types to the orogen fundamental types in orogen

(And, my own wishlist)
 - use the orogen loaders in orogen's codegen part (it's only used in the spec part)
 - replace the absolutely terrible opaque handling in orogen by using type metadata (that would lead to a major simplication of the typekit codebase)

**Comment by marvin2k** [comment]
The gccxml importer does call `add_standard_cxx_types`, removing the call leads to fails due to missing `/std/string`.

The clang importer does not call `add_standard_cxx_types` and circumvents the `/std/string` problem by manually [renaming](https://github.com/orocos-toolchain/typelib/blob/master/tools/typelib-clang-tlb-importer/IgnoredOrRenamedTypes.cpp#L73) `/std/basic_string` to `/std/string`

> In the long run, I'd like to even move on the orogen side to something closer to the ROS convention (no _t and float32/float64 for float and double)

Hm... would add a lot of additional frictional losses...

**Comment by doudou** [comment]
> Hm... would add a lot of additional frictional losses...

Typelib / orogen and so on is *not all about C++* *(anymore). It started that way, but I now passed that phase for the last 5 years or so. 

While I still technically believe that the current approach is the better one, using an IDL has this one major advantage: making people think less about the programming language they use and more about their datatypes / the overall system integration problem.

float32/float64 (just as (u)intXX) are much better descriptions of what the types actually are, and avoid this issue of thinking always in terms of one programming language.

**Comment by doudou** [comment]
OK ... since this seems to make things better and is indeed a bugfix, I merge

**Comment by marvin2k** [comment]
jo.
